File: Sources/App/Classes/Preferences/TPCResourceManager.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2010 - 2018 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import <UniformTypeIdentifiers/UniformTypeIdentifiers.h>

#import "TDCAlert.h"
#import "TPCApplicationInfo.h"
#import "TPCPathInfo.h"
#import "TLOLocalization.h"
#import "TPCResourceManagerPrivate.h"

NS_ASSUME_NONNULL_BEGIN

NSString *const TPCResourceManagerBundleDocumentTypeExtension = @".bundle";
NSString *const TPCResourceManagerBundleDocumentTypeExtensionWithoutPeriod = @"bundle";

NSString *const TPCResourceManagerScriptDocumentTypeExtension = @".scpt";
NSString *const TPCResourceManagerScriptDocumentTypeExtensionWithoutPeriod = @"scpt";

@implementation TPCResourceManager

+ (void)copyResourcesToApplicationSupportFolder
{
	/* Add a system link for the unsupervised scripts folder if it exists. */
	NSString *sourcePath = [TPCPathInfo customScripts];

	NSString *destinationPath =
		[[TPCPathInfo groupContainerApplicationSupport] stringByAppendingPathComponent:@"/Custom Scripts/"];

	if ([RZFileManager() fileExistsAtPath:sourcePath] && [RZFileManager() fileExistsAtPath:destinationPath] == NO) {
		[RZFileManager() createSymbolicLinkAtPath:destinationPath withDestinationPath:sourcePath error:NULL];
	}
}

+ (NSCache *)sharedResourcesCache
{
	static NSCache *cachedValue = nil;

	static dispatch_once_t onceToken;

	dispatch_once(&onceToken, ^{
		cachedValue = [NSCache new];
	});

	return cachedValue;
}

+ (nullable NSDictionary<NSString *, id> *)dictionaryFromResources:(NSString *)name
{
	return [self dictionaryFromResources:name inDirectory:nil key:nil cacheValue:YES];
}

+ (nullable NSDictionary<NSString *, id> *)dictionaryFromResources:(NSString *)name
													   inDirectory:(nullable NSString *)subpath
{
	return [self dictionaryFromResources:name inDirectory:subpath key:nil cacheValue:YES];
}

+ (nullable NSDictionary<NSString *, id> *)dictionaryFromResources:(NSString *)name
													   inDirectory:(nullable NSString *)subpath
														cacheValue:(BOOL)cacheValue
{
	return [self dictionaryFromResources:name inDirectory:subpath key:nil cacheValue:cacheValue];
}

+ (nullable NSDictionary<NSString *, id> *)dictionaryFromResources:(NSString *)name cacheValue:(BOOL)cacheValue
{
	return [self dictionaryFromResources:name inDirectory:nil key:nil cacheValue:cacheValue];
}

+ (nullable NSDictionary<NSString *, id> *)dictionaryFromResources:(NSString *)name key:(nullable NSString *)key
{
	return [self dictionaryFromResources:name inDirectory:nil key:key cacheValue:YES];
}

+ (nullable NSDictionary<NSString *, id> *)dictionaryFromResources:(NSString *)name
															   key:(nullable NSString *)key
														cacheValue:(BOOL)cacheValue
{
	return [self dictionaryFromResources:name inDirectory:nil key:key cacheValue:cacheValue];
}

+ (nullable NSDictionary<NSString *, id> *)dictionaryFromResources:(NSString *)name
													   inDirectory:(nullable NSString *)subpath
															   key:(nullable NSString *)key
{
	return [self dictionaryFromResources:name inDirectory:subpath key:key cacheValue:YES];
}

+ (nullable NSDictionary<NSString *, id> *)dictionaryFromResources:(NSString *)name
													   inDirectory:(nullable NSString *)subpath
															   key:(nullable NSString *)key
														cacheValue:(BOOL)cacheValue
{
	return [self objectFromResources:name
						 inDirectory:subpath
								 key:key
							  kindOf:[NSDictionary class]
						  cacheValue:cacheValue];
}

+ (nullable NSArray *)arrayFromResources:(NSString *)name
{
	return [self arrayFromResources:name inDirectory:nil key:nil cacheValue:YES];
}

+ (nullable NSArray *)arrayFromResources:(NSString *)name inDirectory:(nullable NSString *)subpath
{
	return [self arrayFromResources:name inDirectory:subpath key:nil cacheValue:YES];
}

+ (nullable NSArray *)arrayFromResources:(NSString *)name
							 inDirectory:(nullable NSString *)subpath
							  cacheValue:(BOOL)cacheValue
{
	return [self arrayFromResources:name inDirectory:subpath key:nil cacheValue:cacheValue];
}

+ (nullable NSArray *)arrayFromResources:(NSString *)name cacheValue:(BOOL)cacheValue
{
	return [self arrayFromResources:name inDirectory:nil key:nil cacheValue:cacheValue];
}

+ (nullable NSArray *)arrayFromResources:(NSString *)name key:(nullable NSString *)key
{
	return [self arrayFromResources:name inDirectory:nil key:key cacheValue:YES];
}

+ (nullable NSArray *)arrayFromResources:(NSString *)name key:(nullable NSString *)key cacheValue:(BOOL)cacheValue
{
	return [self arrayFromResources:name inDirectory:nil key:key cacheValue:cacheValue];
}

+ (nullable NSArray *)arrayFromResources:(NSString *)name
							 inDirectory:(nullable NSString *)subpath
									 key:(nullable NSString *)key
{
	return [self arrayFromResources:name inDirectory:subpath key:key cacheValue:YES];
}

+ (nullable NSArray *)arrayFromResources:(NSString *)name
							 inDirectory:(nullable NSString *)subpath
									 key:(nullable NSString *)key
							  cacheValue:(BOOL)cacheValue
{
	return [self objectFromResources:name inDirectory:subpath key:key kindOf:[NSArray class] cacheValue:cacheValue];
}

+ (nullable id)objectFromResources:(NSString *)name
					   inDirectory:(nullable NSString *)subpath
							   key:(nullable NSString *)key
							kindOf:(Class)class
						cacheValue:(BOOL)cacheValue
{
	NSParameterAssert(name != nil);

	/* Bypass cache? */
	if (cacheValue == NO) {
		return [self _objectFromResources:name inDirectory:subpath key:key kindOf:class];
	}

	/* Cache hit? */
	NSString *cacheKey = [NSString
		stringWithFormat:@"%@.plist / %@ / %@", name, ((subpath) ?: @"Root Folder"), ((key) ?: @"Root Object")];

	NSCache *cache = self.sharedResourcesCache;

	id cachedValue = [cache objectForKey:cacheKey];

	if (cachedValue) {
		return cachedValue;
	}

	/* No cache hit, generate */
	cachedValue = [self _objectFromResources:name inDirectory:subpath key:key kindOf:class];

	if (cachedValue) {
		[cache setObject:cachedValue forKey:cacheKey];
	}

	return cachedValue;
}

+ (nullable id)_objectFromResources:(NSString *)name
						inDirectory:(nullable NSString *)subpath
								key:(nullable NSString *)key
							 kindOf:(Class)class
{
	/* Locate resource */
	NSURL *resourceURL = [RZMainBundle() URLForResource:name withExtension:@"plist" subdirectory:subpath];

	if (resourceURL == nil) {
		LogToConsoleError(
			"Resource '%{public}@' in subpath '%{public}@' was not found.", name, ((subpath) ?: @"<No subpath>"));

		return nil;
	}

	/* Read resource */
	NSError *readError = nil;

	NSData *fileContents = [NSData dataWithContentsOfURL:resourceURL options:0 error:&readError];

	if (readError) {
		LogToConsoleError("Resource '%{public}@' could not be read with error: %{public}@",
						  resourceURL.standardizedTildePath,
						  readError.localizedDescription);

		return nil;
	}

	/* Process as a property list */
	NSError *parseError = nil;

	id propertyList = [NSPropertyListSerialization propertyListWithData:fileContents
																options:NSPropertyListImmutable
																 format:NULL
																  error:&parseError];

	if (parseError) {
		LogToConsoleFault("Resource '%{public}@' could not be parsed as a property list with error: %{public}@",
						  resourceURL.standardizedTildePath,
						  parseError.localizedDescription);

		return nil;
	}

	/* Locate object */
	/* Until we know otherwise, the value is the root object. */
	id objectValue = nil;

	if (key == nil) {
		objectValue = propertyList;
	} else {
		/* Property list can be an array. */
		if ([propertyList isKindOfClass:[NSDictionary class]] == NO) {
			LogToConsoleError("Contents of resource '%{public}@' is not a dictionary. "
							  "Cannot locate value of 'key' in other formats.",
							  resourceURL.standardizedTildePath);

			return nil;
		}

		objectValue = [propertyList objectForKey:key];
	}

	if ([objectValue isKindOfClass:class] == NO) {
		LogToConsoleError("Contents of key '%{public}@' in resource '%{public}@' is not kind of class: %{public}@",
						  ((key) ?: @"<Root Object>"),
						  resourceURL.standardizedTildePath,
						  NSStringFromClass(class));

		return nil;
	}

	return objectValue;
}

@end

#pragma mark -

@implementation TPCResourceManagerDocumentTypeImporter

+ (BOOL)autosavesInPlace
{
	/* We are read-only. This suppresses a warning in console. */
	return YES;
}

/* These match CFBundleTypeName of the document types declared in Info.plist */
#define _scriptDocumentTypeName @"Glasstual IRC Client Script"
#define _pluginDocumentTypeName @"Glasstual IRC Client Extension"

- (BOOL)readFromURL:(NSURL *)url ofType:(NSString *)typeName error:(NSError **)outError
{
	/* The document type name handed to us by NSDocumentController is
	 authoritative. The uniform type of the file is consulted as a
	 fallback so that the file's actual type, not its path, decides. */
	if ([typeName isEqualToString:_scriptDocumentTypeName]) {
		[self performImportOfScriptFile:url];

		return YES;
	}

	if ([typeName isEqualToString:_pluginDocumentTypeName]) {
		[self performImportOfPluginFile:url];

		return YES;
	}

	UTType *contentType = nil;

	[url getResourceValue:&contentType forKey:NSURLContentTypeKey error:NULL];

	if (contentType == nil) {
		contentType = [UTType typeWithFilenameExtension:url.pathExtension];
	}

	if (contentType == nil) {
		return NO;
	}

	UTType *scriptType = [UTType typeWithFilenameExtension:TPCResourceManagerScriptDocumentTypeExtensionWithoutPeriod];

	if (scriptType && [contentType conformsToType:scriptType]) {
		[self performImportOfScriptFile:url];

		return YES;
	}

	if ([contentType conformsToType:UTTypeBundle] &&
		[url.pathExtension isEqualToString:TPCResourceManagerBundleDocumentTypeExtensionWithoutPeriod]) {
		[self performImportOfPluginFile:url];

		return YES;
	}

	return NO;
}

#undef _scriptDocumentTypeName
#undef _pluginDocumentTypeName

#pragma mark -
#pragma mark Custom Plugin Files

- (void)performImportOfPluginFile:(NSURL *)url
{
	NSParameterAssert(url != nil);

	NSString *filename = url.lastPathComponent;

	BOOL performInstall = [TDCAlert modalAlertWithMessage:TXTLS(@"Prompts[6tj-yp]")
													title:TXTLS(@"Prompts[xfl-8e]", filename)
											defaultButton:TXTLS(@"Prompts[mvh-ms]")
										  alternateButton:TXTLS(@"Prompts[99q-gg]")];

	if (performInstall == NO) {
		return;
	}

	NSURL *newPath = [[TPCPathInfo customExtensionsURL] URLByAppendingPathComponent:filename];

	BOOL didImport = [self import:url into:newPath];

	if (didImport) {
		NSString *filenameWithoutExtension = filename.stringByDeletingPathExtension;

		[TDCAlert modalAlertWithMessage:TXTLS(@"Prompts[k69-q0]")
								  title:TXTLS(@"Prompts[xek-0t]", filenameWithoutExtension)
						  defaultButton:TXTLS(@"Prompts[c7s-dq]")
						alternateButton:nil];
	}
}

#pragma mark -
#pragma mark Custom Script Files

- (BOOL)panel:(id)sender validateURL:(NSURL *)url error:(NSError **)outError
{
	NSString *scriptsPath = [TPCPathInfo customScripts];

	if ([url.path hasPrefix:scriptsPath] == NO) {
		if (outError) {
			NSMutableDictionary<NSString *, id> *userInfo = [NSMutableDictionary dictionary];

			userInfo[NSURLErrorKey] = url;

			userInfo[NSLocalizedDescriptionKey] = TXTLS(@"Prompts[m2r-gv]");
			userInfo[NSLocalizedRecoverySuggestionErrorKey] = TXTLS(@"Prompts[ztu-nv]");

			*outError = [NSError errorWithDomain:TXErrorDomain code:27984 userInfo:userInfo];
		}

		return NO;
	}

	return YES;
}

- (void)performImportOfScriptFile:(NSURL *)url
{
	NSParameterAssert(url != nil);

	NSString *filename = url.lastPathComponent;

	/* Ask user before installing. */
	BOOL performInstall = [TDCAlert modalAlertWithMessage:TXTLS(@"Prompts[6tj-yp]")
													title:TXTLS(@"Prompts[xfl-8e]", filename)
											defaultButton:TXTLS(@"Prompts[mvh-ms]")
										  alternateButton:TXTLS(@"Prompts[99q-gg]")];

	if (performInstall == NO) {
		return; // Do not install.
	}

	NSURL *folderRep = [TPCPathInfo customScriptsURL];

	if ([RZFileManager() fileExistsAtURL:folderRep] == NO) {
		folderRep = [TPCPathInfo userApplicationScriptsURL];
	}

	NSString *bundleID = [TPCApplicationInfo applicationBundleIdentifier];

	NSSavePanel *d = [NSSavePanel savePanel];

	d.delegate = (id)self;

	d.canCreateDirectories = YES;

	d.directoryURL = folderRep;

	d.title = TXTLS(@"Prompts[6hx-ni]");

	d.message = TXTLS(@"Prompts[0bj-ic]", bundleID);

	d.nameFieldStringValue = url.lastPathComponent;

	d.showsTagField = NO;

	[d beginWithCompletionHandler:^(NSInteger returnCode) {
		if (returnCode == NSModalResponseOK) {
			if ([self import:url into:d.URL] == NO) {
				return;
			}

			NSString *filename = d.URL.lastPathComponent;

			XRPerformBlockAsynchronouslyOnMainQueue(^{
				[self performImportOfScriptFilePostflight:filename];
			});
		}
	}];
}

- (void)performImportOfScriptFilePostflight:(NSString *)filename
{
	NSParameterAssert(filename != nil);

	NSString *filenameWithoutExtension = filename.stringByDeletingPathExtension;

	[TDCAlert modalAlertWithMessage:TXTLS(@"Prompts[3ze-xh]", filenameWithoutExtension)
							  title:TXTLS(@"Prompts[4ua-v5]", filenameWithoutExtension)
					  defaultButton:TXTLS(@"Prompts[c7s-dq]")
					alternateButton:nil];
}

#pragma mark -
#pragma mark General Import Controller

- (BOOL)import:(NSURL *)url into:(NSURL *)destination
{
	return [RZFileManager() replaceItemAtURL:destination
							   withItemAtURL:url
									 options:(CSFileManagerOptionsMoveToTrash | CSFileManagerOptionsRemoveIfExists)];
}

@end

NS_ASSUME_NONNULL_END

File: Sources/App/Modules/ApplicationSupport/TPCResourceManager.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \\ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\\___/_/\\_\\__|\\__,_|\\__,_|_|
 *
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 *********************************************************************** */

import CocoaExtensions
import Foundation
import os

public nonisolated enum ResourceDocumentType {
	public static let bundleFileExtension = ".bundle"
	public static let bundleFilenameExtension = "bundle"
	public static let scriptFileExtension = ".scpt"
	public static let scriptFilenameExtension = "scpt"
}

@objc(TPCResourceManager)
public final nonisolated class ResourceManager: NSObject {
	private static let logger = Logger(
		subsystem: Bundle.main.bundleIdentifier ?? "Glasstual",
		category: "ResourceManager"
	)

	private nonisolated(unsafe) static let resourcesCache = NSCache<AnyObject, AnyObject>()

	@objc public static var sharedResourcesCache: NSCache<AnyObject, AnyObject> {
		resourcesCache
	}

	@objc public static func copyResourcesToApplicationSupportFolder() {
		guard let sourcePath = PathInfo.customScripts,
		      let destinationRoot = PathInfo.groupContainerApplicationSupport
		else {
			return
		}

		let destinationPath = (destinationRoot as NSString).appendingPathComponent("/Custom Scripts/")
		let fileManager = FileManager.default

		guard fileManager.fileExists(atPath: sourcePath),
		      fileManager.fileExists(atPath: destinationPath) == false
		else {
			return
		}

		try? fileManager.createSymbolicLink(atPath: destinationPath, withDestinationPath: sourcePath)
	}

	// MARK: - Dictionaries

	@objc(dictionaryFromResources:)
	public static func dictionary(fromResources name: String) -> [String: Any]? {
		dictionary(fromResources: name, inDirectory: nil, key: nil, cacheValue: true)
	}

	@objc(dictionaryFromResources:cacheValue:)
	public static func dictionary(fromResources name: String, cacheValue: Bool) -> [String: Any]? {
		dictionary(fromResources: name, inDirectory: nil, key: nil, cacheValue: cacheValue)
	}

	@objc(dictionaryFromResources:inDirectory:)
	public static func dictionary(fromResources name: String, inDirectory subpath: String?) -> [String: Any]? {
		dictionary(fromResources: name, inDirectory: subpath, key: nil, cacheValue: true)
	}

	@objc(dictionaryFromResources:inDirectory:cacheValue:)
	public static func dictionary(
		fromResources name: String,
		inDirectory subpath: String?,
		cacheValue: Bool
	) -> [String: Any]? {
		dictionary(fromResources: name, inDirectory: subpath, key: nil, cacheValue: cacheValue)
	}

	@objc(dictionaryFromResources:key:)
	public static func dictionary(fromResources name: String, key: String?) -> [String: Any]? {
		dictionary(fromResources: name, inDirectory: nil, key: key, cacheValue: true)
	}

	@objc(dictionaryFromResources:key:cacheValue:)
	public static func dictionary(fromResources name: String, key: String?, cacheValue: Bool) -> [String: Any]? {
		dictionary(fromResources: name, inDirectory: nil, key: key, cacheValue: cacheValue)
	}

	@objc(dictionaryFromResources:inDirectory:key:)
	public static func dictionary(
		fromResources name: String,
		inDirectory subpath: String?,
		key: String?
	) -> [String: Any]? {
		dictionary(fromResources: name, inDirectory: subpath, key: key, cacheValue: true)
	}

	@objc(dictionaryFromResources:inDirectory:key:cacheValue:)
	public static func dictionary(
		fromResources name: String,
		inDirectory subpath: String?,
		key: String?,
		cacheValue: Bool
	) -> [String: Any]? {
		object(fromResources: name, inDirectory: subpath, key: key, kindOf: NSDictionary.self, cacheValue: cacheValue)
			as? [String: Any]
	}

	// MARK: - Arrays

	@objc(arrayFromResources:)
	public static func array(fromResources name: String) -> [Any]? {
		array(fromResources: name, inDirectory: nil, key: nil, cacheValue: true)
	}

	@objc(arrayFromResources:cacheValue:)
	public static func array(fromResources name: String, cacheValue: Bool) -> [Any]? {
		array(fromResources: name, inDirectory: nil, key: nil, cacheValue: cacheValue)
	}

	@objc(arrayFromResources:inDirectory:)
	public static func array(fromResources name: String, inDirectory subpath: String?) -> [Any]? {
		array(fromResources: name, inDirectory: subpath, key: nil, cacheValue: true)
	}

	@objc(arrayFromResources:inDirectory:cacheValue:)
	public static func array(fromResources name: String, inDirectory subpath: String?, cacheValue: Bool) -> [Any]? {
		array(fromResources: name, inDirectory: subpath, key: nil, cacheValue: cacheValue)
	}

	@objc(arrayFromResources:key:)
	public static func array(fromResources name: String, key: String?) -> [Any]? {
		array(fromResources: name, inDirectory: nil, key: key, cacheValue: true)
	}

	@objc(arrayFromResources:key:cacheValue:)
	public static func array(fromResources name: String, key: String?, cacheValue: Bool) -> [Any]? {
		array(fromResources: name, inDirectory: nil, key: key, cacheValue: cacheValue)
	}

	@objc(arrayFromResources:inDirectory:key:)
	public static func array(fromResources name: String, inDirectory subpath: String?, key: String?) -> [Any]? {
		array(fromResources: name, inDirectory: subpath, key: key, cacheValue: true)
	}

	@objc(arrayFromResources:inDirectory:key:cacheValue:)
	public static func array(
		fromResources name: String,
		inDirectory subpath: String?,
		key: String?,
		cacheValue: Bool
	) -> [Any]? {
		object(fromResources: name, inDirectory: subpath, key: key, kindOf: NSArray.self, cacheValue: cacheValue)
			as? [Any]
	}

	// MARK: - Generic object loading

	@objc(objectFromResources:inDirectory:key:kindOf:cacheValue:)
	public static func object(
		fromResources name: String,
		inDirectory subpath: String?,
		key: String?,
		kindOf type: AnyClass,
		cacheValue: Bool
	) -> Any? {
		if cacheValue == false {
			return loadObject(fromResources: name, inDirectory: subpath, key: key, kindOf: type)
		}

		let cacheKey = cacheKey(name: name, subpath: subpath, key: key) as NSString

		if let cachedValue = resourcesCache.object(forKey: cacheKey) {
			return cachedValue
		}

		guard let loadedValue = loadObject(fromResources: name, inDirectory: subpath, key: key, kindOf: type) else {
			return nil
		}

		resourcesCache.setObject(loadedValue as AnyObject, forKey: cacheKey as AnyObject)

		return loadedValue
	}

	private static func loadObject(
		fromResources name: String,
		inDirectory subpath: String?,
		key: String?,
		kindOf type: AnyClass
	) -> Any? {
		guard let resourceURL = Bundle.main.url(forResource: name, withExtension: "plist", subdirectory: subpath) else {
			logger.error(
				"Resource '\(name, privacy: .public)' in subpath '\(subpath ?? "<No subpath>", privacy: .public)' was not found."
			)

			return nil
		}

		let fileContents: Data

		do {
			fileContents = try Data(contentsOf: resourceURL)
		} catch {
			logger.error(
				"Resource '\(Self.displayPath(for: resourceURL), privacy: .public)' could not be read with error: \(error.localizedDescription, privacy: .public)"
			)

			return nil
		}

		let propertyList: Any

		do {
			propertyList = try PropertyListSerialization.propertyList(from: fileContents, options: [], format: nil)
		} catch {
			logger.fault(
				"Resource '\(Self.displayPath(for: resourceURL), privacy: .public)' could not be parsed as a property list with error: \(error.localizedDescription, privacy: .public)"
			)

			return nil
		}

		let objectValue: Any?

		if let key {
			guard let dictionary = propertyList as? [String: Any] else {
				logger.error(
					"Contents of resource '\(Self.displayPath(for: resourceURL), privacy: .public)' is not a dictionary. Cannot locate value of 'key' in other formats."
				)

				return nil
			}

			objectValue = dictionary[key]
		} else {
			objectValue = propertyList
		}

		guard let objectValue, (objectValue as AnyObject).isKind(of: type) else {
			logger.error(
				"Contents of key '\(key ?? "<Root Object>", privacy: .public)' in resource '\(Self.displayPath(for: resourceURL), privacy: .public)' is not kind of class: \(NSStringFromClass(type), privacy: .public)"
			)

			return nil
		}

		return objectValue
	}

	private static func cacheKey(name: String, subpath: String?, key: String?) -> String {
		"\(name).plist / \(subpath ?? "Root Folder") / \(key ?? "Root Object")"
	}

	private static func displayPath(for url: URL) -> String {
		(url as NSURL).textualStandardizedTildePath ?? url.path
	}
}
